### ISSUE TITLE
Product : Retrieving related object with single request

### ISSUE THREAD

**Issue by RomainJufer** [DefectDojo/django-DefectDojo#3018]
This proposal has been made conjointly with @xens 

We would like to reduce the number of queries required to fetch all the data related to a `Product`. More specifically, we would like to provide a way to retrieve in a single query the data associated with fields defining a relation.

To do that we came up with essentially three proposals to achieve this :

1. Directly modifying the ProductSerializer to force it to follows these relations and get the data. The django framework support this feature natively and thus it is really easy to implement. However, the downside is that it would mean forcing a complete retrieval even if it is not needed.

2. Modifying the endpoint to support an additional parameter specifying if we want standard (as currently implemented) or full retrieval. Something along the line of `/products/?full`.

3. Modifying the endpoints to be even more parametric and be able to specify for which field we want to follow the relation.

As an example, we have implemented the retrieval for `product_manager` yielding

```
{
      "id": 3,
      "findings_count": 0,
      "findings_list": [],
      "product_manager": {
        "id": 1,
        "username": "admin",
        "first_name": "Admin",
        "last_name": "User",
        "email": "admin@defectdojo.local",
        "last_login": "2020-10-16T07:31:49.047121Z",
        "is_active": true,
        "is_staff": true
      },
      "tags": [],
      "product_meta": [],
      ...
}
```

instead of 

```
{
      "id": 3,
      "findings_count": 0,
      "findings_list": [],
      "product_manager": 5,
      "tags": [],
      "product_meta": [],
      ...
}
```

**Comment by valentijnscholten** [comment]
How does option 1 work? Could it be used but only when a parameter is specified (option 2). If possible, you don't want to write all the code to get all the relations (and their relations etc). How many levels deep will it go?

**Comment by RomainJufer** [comment]
For option 1 we simply modify the `ProductSerializer` using the serializers corresponding to the fields for which we want to follow the relation and load the data. For example, for `product_manager` we use the `UserSerializer`. In our case we only need to follow the relation for the fields : 

- product_manager
- technical_manager
- team_manager
- prod_type
- regulations

We would like to avoid doing it for `findings` as the list can be relatively big. This is why we also propose option 3 that offers more granularity and allows us to generalize the approach. While option 2 is also possible we think that it would not scale well if, for example, we include also `findings` which might be required to be semantically coherent. For the depths of the retrieval, we only need 1 level in our case. I dont know if this is something worth adding as a parameter and also how easy it would be to implement it.

Let me know if my explanation is not clear or if I did not answer your questions :-)

**Comment by valentijnscholten** [comment]
yeah 3 is best and maybe not too much work if you can reuse the serializers for those relations.

**Comment by RomainJufer** [comment]
I  did the implementation for the field `technical_manager` to illustrate the changes. This amount to adding a few lines of code inside the `ProductSerializer` as the snippet below shows : 

```
 def to_representation(self, instance):
        parameters = self.context["request"].GET
        rep = super().to_representation(instance)

        if "technical_contact" in parameters and not rep["technical_contact"] is None:
            rep.pop("technical_contact")
            rep["technical_contact"] = UserSerializer(read_only=True).to_representation(instance.technical_contact)
        return rep
```

Then the endpoint for getting products can be called with the additionnal parameter `technical_manager`. 

This is the most elegant way I found of doing it but since I'm not a Django expert, feel free to let me know if I should do it differently. Also, for the parameters, if you think that we should use different names than the names of the fields we want to expand, let me know as well.

**Comment by valentijnscholten** [comment]
I think it's elegant enough. It's all glue code, maybe it can even be made more generic doing something like `for relation in ['technical_contact', 'product_manager', ...]: <the code above>`

for the parameters I was thinking maybe it's nice to have some comma seperated parameter, i.e. `include_related=technical_contact, product_manager, ...`.  Or multiple of the same, i.e. ìnclude=technical_contact&include=product_manager....`

**Comment by RomainJufer** [comment]
We would like to get back on this proposal with some improvements (I've close the related PR).

**Context :** We have noticed that the feature that we proposed for the `Product` would be needed for other endpoint as well (`Finding` to name one). Therefore, we decided to see if this feature was generizable.

**Observations :** As can be seen in the code above, we implemented the prefetching directly inside the serializers. While this was initially a good approach it has a few downsides. 

- It induces more burden on the caller as, depending on the parameters given to `prefetch`, it produces JSON with fields that can either store an integer (the primary key) or an object. Therefore, unmarshalling of the response is more difficult and might be even impossible if the programming language used by the client does not support some kind of union type.
- The auto generated swagger documentation is wrong as the existing inspectors cannot derive both the newly added query parameter nor support union type in the response model.
- The data is added per product even if it's redundant which could yield large response

Thus, we decided to see if by generalizing we could also solve these problems.

**Solution :** We decided to implement prefetching by creating new `mixin` that implement the basic `read` and `list` operations so that it can be easily deployed on the selected endpoint. We take the example of the new list mixin as the read mixin is completly analogous.

Our list mixin, `class PrefetchListMixin(mixins.ListModelMixin)` only override the `list(...)` operation. The method is similar to the original one with the difference that it keep the prefetched data as a set. The result produced, for findings for example, is
```
{
   count: ...,
   previous: ...,
   next: ...,
   results: [ /* findings */],
   prefetch: {
       field_name1: {
           id1: { /* object1 */ },
           id2: { /* object2 */ }
      },
      field_name2: { ... },
      ...
   }
```

The main advantage with this method is that we do not have the typing problem mentionned above and also we avoid duplicated data. The `prefetch` field acts as a form of cache as it was intended.

The prefetching mechanism is completly automatized. The process goes according to the following (simplied) algorithm.
```
input: 
    - object: ModelInstance // the object retrieved from the database
    - params: list[string] // the field we want to prefetch
output: 
    - prefetched: dict()

prefetched = dict()
 // we gather all the available serializers in a map type -> serializer_class where type is the type of the model parsed by the serializer_class is the type of the serializer
serializers = build_serializers()

for param in params:
    // Find the best serializer that can be applied
    serializer = match_serializer(object[param], serializers)
    if serializer is None:
         continue
    data = serializer().to_representation(object[param])
    prefetched[param][data.id] = data

return prefetched
```

This process has been implemented and works well. For an endpoint represented as a `ViewSet` we only need to add the new mixin to support the operation. However, this does not resolve the problem of the swagger documentation. In our case, we need to add the new parameter with the list of field that can be prefetched as well as the new field in the response. The best way we found to do it is first to extand the `AutoSwaggerSchema` class to support the manual addition of new parameters and field inside the response. Then we provide a new decorator that can be added to a ViewSet and that compute automatically these new information without requiring any manual manipulation.

In the end, supporting prefetching for an endpoint can be done by writing (for example for the `Product` endpoints)
```
@prefetch_decorator(name="list", serializer=ProductSerializer)
class ProductViewSet(PrefetchListMixin, /* existing mixing */):
        /* original code */
```

We do you think ?

**Comment by valentijnscholten** [comment]
Sounds interesting, but would need to see the code to fully understand it.

**Comment by valentijnscholten** [comment]
Implemented in #3516
